File: ruby/linked_list_sum.rb
class Node
  attr_accessor :val, :next

  def initialize(val)
    @val = val
    @next = nil
  end
end

# def linked_list_sum(head)
#   sum = 0 
#   current = head 
#   while !current.nil? do 
#     sum += current.val 
#     current = current.next
#   end
#   sum 
# end

def linked_list_sum(head)
  return 0 if head.nil?
  head.val + linked_list_sum(head.next)
end

a = Node.new(1)
b = Node.new(2)
c = Node.new(3)
d = Node.new(4)

a.next = b
b.next = c
c.next = d

p linked_list_sum(a)

File: python/linked_list_sum.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

# def linked_list_sum(head):
#     sum = 0 
#     current = head
#     while current is not None:
#         sum += current.val
#         current = current.next
#     return sum 

def sum_list(head):
  if head is None:
    return 0
  return head.val + sum_list(head.next)